src/Servers/HttpSys/src/RequestProcessing/Response.cs

```diff
@@ -527,111 +527,89 @@ private static bool Matches(string knownValue, StringValues input)
-                        }
+                        // Add Name
+                        bytes = HeaderEncoding.GetBytes(headerName, writer);
+                        unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].NameLength = (ushort)bytes.Length;
+                        unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].pName = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(bytes));
+
+                        // Add Value
+                        headerValue = headerValues[headerValueIndex] ?? string.Empty;
+                        bytes = HeaderEncoding.GetBytes(headerValue, writer);
+                        unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].RawValueLength = (ushort)bytes.Length;
+                        unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].pRawValue = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(bytes));
+                        _nativeResponse.Response_V1.Headers.UnknownHeaderCount++;
                     }
-                    else if (headerPair.Value.Count == 1)
+                }
+                else if (headerPair.Value.Count == 1)
+                {
+                    headerValue = headerValues[0] ?? string.Empty;
+                    bytes = HeaderEncoding.GetBytes(headerValue, writer);
+                    pKnownHeaders[lookup].RawValueLength = (ushort)bytes.Length;
+                    pKnownHeaders[lookup].pRawValue = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(bytes));
+                }
+                else
+                {
+                    if (knownHeaderInfo == null)
                     {
-                        headerValue = headerValues[0] ?? string.Empty;
-                        bytes = HeaderEncoding.GetBytes(headerValue);
-                        pKnownHeaders[lookup].RawValueLength = (ushort)bytes.Length;
-                        gcHandle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
-                        pinnedHeaders.Add(gcHandle);
-                        pKnownHeaders[lookup].pRawValue = (byte*)gcHandle.AddrOfPinnedObject();
+                        var responseAlloc = (HttpApiTypes.HTTP_RESPONSE_INFO*)writer.Alloc(sizeof(HttpApiTypes.HTTP_RESPONSE_INFO) * numKnownMultiHeaders);
+                        knownHeaderInfo = new Span<HttpApiTypes.HTTP_RESPONSE_INFO>(responseAlloc, numKnownMultiHeaders);
+                        _nativeResponse.pResponseInfo = responseAlloc;
                     }
-                    else
+
+                    knownHeaderInfo[_nativeResponse.ResponseInfoCount].Type = HttpApiTypes.HTTP_RESPONSE_INFO_TYPE.HttpResponseInfoTypeMultipleKnownHeaders;
+                    knownHeaderInfo[_nativeResponse.ResponseInfoCount].Length = (uint)sizeof(HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS);
+
+                    var header = (HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS*)writer.Alloc(sizeof(HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS));
```

Comment:
It would be better if this tried to used memory from the pooled block instead of a separate allocation.